Mathlib/Data/Matrix/Rank.lean

```diff
@@ -172,6 +305,28 @@ theorem rank_diagonal [Fintype m] [DecidableEq m] [DecidableEq R] (w : m → R)
   rw [Matrix.rank, ← Matrix.toLin'_apply', Module.finrank, ← LinearMap.rank,
     LinearMap.rank_diagonal, Cardinal.toNat_natCast]
 
+theorem cRank_diagonal [DecidableEq m] (w : m → R) :
+    (diagonal w).cRank = lift.{uR} #{i // (w i) ≠ 0} := by
+  classical
+  set w' : {i // (w i) ≠ 0} → _ := fun i ↦ (diagonal w) i
+  have h : LinearIndependent R w' := by
+    have hli' := Pi.linearIndependent_single_ne_zero
+      (v := fun i : m ↦ if w i = 0 then (1 : R) else w i) (by simp [ite_eq_iff'])
+    convert hli'.comp Subtype.val Subtype.val_injective
+    ext ⟨j, hj⟩ k
+    simp [w', diagonal, hj, Pi.single_apply, eq_comm]
+  have hrw : insert 0 (range (diagonal w)ᵀ) = insert 0 (range w') := by
+    suffices ∀ a, diagonal w a = 0 ∨ ∃ b, w b ≠ 0 ∧ diagonal w b = diagonal w a
+      by simpa [subset_antisymm_iff, subset_def, w']
+    simp_rw [or_iff_not_imp_right, not_exists, not_and, not_imp_not]
+    simp +contextual [funext_iff, diagonal, or_iff_not_imp_right]
+  rw [cRank, ← span_insert_zero, hrw, span_insert_zero, rank_span h,
+    ← lift_umax, ← Cardinal.mk_range_eq_of_injective h.injective, lift_id']
+
+theorem eRank_diagonal [DecidableEq m] (w : m → R) :
+    (diagonal w).eRank = ENat.card {i // (w i) ≠ 0} := by
```

Comment:
```suggestion
    (diagonal w).eRank = {i | w i ≠ 0}.encard := by
```
is perhaps cleaner

Reply:
Yes, I agree. I've added a simp lemma for `ENat.card {x // P x}` more generally.

Reply:
Update - the @[simp] broke something elsewhere, so I've desimped it in the name of limiting scope.